File: tree-problems.js
import Node from './tree.js'

// initialize trees of random integers from 1 to 100
const nodes = [...Array(11)].map(() => Math.ceil(Math.random() * 100));
const tree1 = new Node(nodes.pop());
for (let n of nodes) {
    tree1.insert(n)
}
const nodes2 = [...Array(11)].map(() => Math.ceil(Math.random() * 100));
const tree2 = new Node(nodes2.pop());
for (let n of nodes2) {
    tree2.insert(n)
}

console.log('tree 1')
tree1.print();
console.log('\ntree 2')
tree2.print();

// IS SAME TREE 
const isSameTree = (p,q) => {
    if (p == null && q == null) return true;
    if (p == null || q == null) return false;
    if (p.value !== q.value) return false;
    return isSameTree(p.left, q.left) && isSameTree(p.right, q.right)
}
console.log('\nIs same tree: tree1 tree1 ' + isSameTree(tree1, tree1));
console.log('Is same tree: tree1 tree2 ' + isSameTree(tree1, tree2));

// MAX DEPTH
const maxDepth = root => {
    if (root == null) return 0;

    let left = maxDepth(root.left);
    let right = maxDepth(root.right)

    let maximum = Math.max(left, right);
    return maximum + 1;
}
console.log('\nMax depth tree 1 is: ' + maxDepth(tree1));
console.log('Max depth tree 2 is: ' + maxDepth(tree2));

// LEVEL ORDER BFS
const levelOrder = root => {
    const result = [];
    const queue = [];

    if (root) {
        queue.push(root)
    }

    while (queue.length) {
        let level = [];
        let count = queue.length;

        for (let i = 0; i < count; i++) {
            let node = queue.shift();
            level.push(node.value);
            if (node.left)
                queue.push(node.left);
            if (node.right)
                queue.push(node.right)
        }
        result.push(level)
    }
    return result;
}
console.log('\nLevel order BFS tree 1');
console.log(levelOrder(tree1));
console.log('Level order BFS tree 2');
console.log(levelOrder(tree2));

File: tree-problems.py
from random import randrange
import tree

# initialize trees of random integers from 1 to 100
nodes1 = [randrange(1,101) for i in range(11)]
tree1 = tree.Node(nodes1.pop())
for n in nodes1:
    tree1.insert(n)
tree1.inorder()

nodes2 = [randrange(1,101) for i in range(11)]
tree2 = tree.Node(nodes2.pop())
for n in nodes2:
    tree2.insert(n)

# IS SAME TREE
def isSameTree(p,q):
    if p == None and q == None:
        return True
    if p == None or q == None:
        return False
    if p.value != q.value:
        return False

    return isSameTree(p.left, q.left) and isSameTree(p.right, q.right)

print('Is same tree ', isSameTree(tree1, tree1))
print('Is same tree ', isSameTree(tree1, tree2))

# MAX DEPTH
def maxDepth(root):
    if root == None:
        return 0

    left = maxDepth(root.left)
    right = maxDepth(root.right)

    maximum = max(left, right)
    return maximum + 1

print('Max depth tree 1 is:', maxDepth(tree1))
print('Max depth tree 2 is:', maxDepth(tree2))

# LEVEL ORDER BFS
def levelOrder(root):
    result = []
    queue = []

    if root:
       queue.append(root)

    while len(queue):
        level = []
        count = len(queue) 

        for i in range(count):
            node = queue.pop(0)
            level.append(node.value)
            if node.left:
                queue.append(node.left)
            if node.right:
                queue.append(node.right)

        result.append(level)

    return result

print('Level order BFS\n', levelOrder(tree1))
